### PR TITLE
feat: implement basic Blockchain, BlockMiner, and MemPool

### PR DESCRIPTION
**Pull request by Wodann**

### CODE REVIEW

File: packages/hardhat-core/test/internal/hardhat-network/helpers/assertions.ts
Comment (Wodann) [inline packages/hardhat-core/test/internal/hardhat-network/helpers/assertions.ts:57]: @fvictorio do we provide guarantees for error messages w.r.t capitalisation of characters?

**Review by feuGeneA — CHANGES_REQUESTED** [review]
Review still in progress, but these are the comments I have so far.

**Review by fvictorio — APPROVED** [review]
I only reviewed the typescript code, but if there's something on the rust side you would like me to check, let me know.

File: crates/rethnet_evm/src/blockchain/local.rs
Comment (Wodann) [inline crates/rethnet_evm/src/blockchain/local.rs:210]: @fvictorio, @alcuadrado This question was intended for you. Theoretically block numbers can be larger than `u64::MAX`, but the amount of memory that can be addresses on a computer cannot. Hence, my question: will we ever need to support block numbers that are larger?

File: crates/rethnet_evm/src/blockchain/forked.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/forked.rs:25]: This error is a direct copy of the Hardhat error message, hence why I'm opting not to add the word "only"

File: crates/rethnet_evm/src/blockchain/local.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/local.rs:210]: The question was directed at the reviewers and should be solved (and removed) before merging). I @'ed Franco & Pato

File: crates/rethnet_evm/src/blockchain/remote.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/remote.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/1b0a93b7ecea339e84905e861b4b6068c7951438

A pending block does not contain several fields, including `total_difficulty`. Since this is guaranteed to not be a pending block, we can call `expect` on the `Option`.

File: crates/rethnet_evm/src/blockchain/remote.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/remote.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/1b0a93b7ecea339e84905e861b4b6068c7951438

A pending block does not contain several fields, including `total_difficulty`. Since this is guaranteed to not be a pending block, we can call `expect` on the `Option`.

File: crates/rethnet_evm/src/blockchain/remote.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/remote.rs:123]: A pending block does not contain several fields, which would result in a failure to convert to a block and its callers. Since this is guaranteed to not be a pending block, we can call `expect` on the `Result`.

File: crates/rethnet_evm/src/blockchain/remote.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/remote.rs:141]: This performs those `N` requests in parallel using async futures. It could be that doing a batch request would be more efficient in terms of bandwidth or runtime performance, but I'd only want to make that change after we do benchmarking - in the future.

File: crates/rethnet_evm/src/blockchain/storage.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/storage.rs:18]: If a hash is duplicated, the number **must** be duplicated too, given that the hash calculation includes the number.

Theoretically, it could be that a block with with a duplicated number has a different hash, but it feels like overkill to have two errors: `DuplicateBlockDueToHash` and `DuplicateBlockDueToNumber`. The error is the fact that the block is duplicated. I will however change the member variable comments to not say "Duplicated" to avoid confusion.

Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/storage.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/bf48a23f39ef5be9e021426a7b1f1a2c0e910aa0

File: crates/rethnet_evm/src/blockchain/local.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/local.rs]: Resolved in https://github.com/NomicFoundation/hardhat/pull/4175/commits/db1b794f82bdc22b8949625262d0d7d2e67f3755

File: crates/rethnet_evm/src/blockchain/forked.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/forked.rs:75]: If we add a new method to a struct/class for any variant of usages for its API it's poor encapsulation. There is a good article in C++ about the meaning of encapsulation: https://drdobbs.com/cpp/how-non-member-functions-improve-encapsu/184401197

This is already executing the 3 requests simulatenously. To optimise the bandwidth, we could convert this into a manual `batch` call, but it comes at the cost of exposing the internals. Alternatively, we could come up with a clever macro that converts any combination of client calls into a single batch request object, but that feels like overkill to spend time on without knowing the performance impact.

File: crates/rethnet_evm/src/blockchain.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain.rs:75]: This is following similarly named functions in Rust collections that use hashing, e.g. https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert

File: crates/rethnet_evm/src/blockchain.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain.rs:63]: I'm opting for `last` here to follow the Rust standards (similar to what I did for RPC client methods). It refers to the last entry using `last`.

I agree, last and latest are synonymous, but `last` also has the benefit of just being shorter.

File: crates/rethnet_evm/benches/state/database_commit.rs
Comment (Wodann) [reply crates/rethnet_evm/benches/state/database_commit.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/ba3d9e522e2fe9e742f357b65ed51fc5c51414cb

File: crates/rethnet_eth/src/remote/eth.rs
Comment (Wodann) [reply crates/rethnet_eth/src/remote/eth.rs]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/9635fff6b697d772e9c9105ae63d4dc0c9be9c92

File: crates/rethnet_evm/src/mempool.rs
Comment (Wodann) [inline crates/rethnet_evm/src/mempool.rs:145]: @fvictorio & @alcuadrado

File: crates/rethnet_evm/src/miner.rs
Comment (Wodann) [inline crates/rethnet_evm/src/miner.rs:156]: We're relying on `revm` to tell us that we exceed the block gas limit. We could check this manually beforehand. What do you think is better @fvictorio & @alcuadrado ?

File: packages/hardhat-core/test/internal/hardhat-network/helpers/assertions.ts
Comment (Wodann) [reply packages/hardhat-core/test/internal/hardhat-network/helpers/assertions.ts:57]: I changed the first letter of a sentence to be capitalised

File: crates/rethnet_evm/src/blockchain/forked.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/forked.rs:269]: Resolved by https://github.com/NomicFoundation/hardhat/pull/4175/commits/1439e9f7bfd0cdc8d4481c999598228bfd5788fd

File: crates/rethnet_evm/src/blockchain/local.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/local.rs]: If you look at the two implementations, you'll notice that there are actually slight differences based on the underlying storage containers, that cannot be de-duplicated. In particular, search based on hash vs search based on index (sparse vs contiguous storage).

I did move the block validation to a shared function. To do so, I added caching of the hash, an optimisation that was needed anyway, to prevent the block hash from being calculated multiple times per call: https://github.com/NomicFoundation/hardhat/pull/4175/commits/a31a065963c6c1a2dc2b31e405829f6608ab59d8

File: crates/rethnet_evm/src/blockchain/local.rs
Comment (alcuadrado) [reply crates/rethnet_evm/src/blockchain/local.rs:203]: We computed it, and it's safe to assume that they feet in a u64, even if blocks were created every second.

File: crates/rethnet_evm/src/mempool.rs
Comment (alcuadrado) [reply crates/rethnet_evm/src/mempool.rs:145]: Every account exists, even if it hasn't been touched and has no state associated to it.

File: crates/rethnet_evm/src/miner.rs
Comment (Wodann) [reply crates/rethnet_evm/src/miner.rs:156]: Don't block on this, but bring it up with Dragan as a possible improvement - by making a strongly typed `TxEnv` that does the validation upon construction.

File: crates/rethnet_evm/src/transaction/pending.rs
Comment (Wodann) [reply crates/rethnet_evm/src/transaction/pending.rs:65]: From the user's perspective, every account exists even if it no state is associated with it, so no need to return an error.

File: crates/rethnet_evm_napi/src/state.rs
Comment (feuGeneA) [inline crates/rethnet_evm_napi/src/state.rs:100]: Could be good to have a comment explaining this line

File: crates/rethnet_evm/src/miner.rs
Comment (Wodann) [reply crates/rethnet_evm/src/miner.rs:16]: All of my use statements are auto-managed by rust-analyzer into these groupings, so I'd not want to have to change the groupings manually for each file.

It doesn't distinguish between first-party and third-party dependencies AFAIK

File: crates/rethnet_evm/src/miner.rs
Comment (Wodann) [reply crates/rethnet_evm/src/miner.rs:157]: A transaction with a lower gas limit might exist. Transactions are ordered in the mempool based on insertion time, not on gas limit.

So even though this transaction is too expensive, an affordable one might exist still

Comment (feuGeneA) [reply crates/rethnet_evm/src/miner.rs:157]: got it, that makes sense, thank you

File: crates/rethnet_evm_napi/src/mempool.rs
Comment (Wodann) [reply crates/rethnet_evm_napi/src/mempool.rs:13]: This is not inside the procedural macro, hence why I'm sticking with the `///` syntactic sugar

File: crates/rethnet_evm/src/miner.rs
Comment (Wodann) [reply crates/rethnet_evm/src/miner.rs:30]: That's a good point. I'm happy to make it full-length generic types to improve readability. I've created an issue to do this for the entire codebase: https://github.com/NomicFoundation/rethnet/issues/146

File: crates/rethnet_evm_napi/src/blockchain.rs
Comment (Wodann) [reply crates/rethnet_evm_napi/src/blockchain.rs:234]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4175/commits/eb051b809524a1ac4c25f10767dca08fd0ec6d7c

File: crates/rethnet_evm_napi/src/state.rs
Comment (Wodann) [reply crates/rethnet_evm_napi/src/state.rs:101]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4175/commits/eb051b809524a1ac4c25f10767dca08fd0ec6d7c

File: crates/rethnet_evm/src/blockchain/storage.rs
Comment (Wodann) [reply crates/rethnet_evm/src/blockchain/storage.rs]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4175/commits/aa7be38e1e37acac4b8fe035c7d4ef3b17cd8d4b by adding punctuation

From a linguistic perspective, it doesn't actually state the the number AND hash are duplicated. The block already exists and these are its hash and number.

File: crates/rethnet_evm/src/transaction/pending.rs
Comment (feuGeneA) [reply crates/rethnet_evm/src/transaction/pending.rs:262]: I see, thanks for teaching me.

**Review by feuGeneA — APPROVED** [review]

**Pull request merged**
